helix-core/src/line_ending.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{Rope, RopeGraphemes, RopeSlice};
+
+/// Represents one of the valid Unicode line endings.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum LineEnding {
+    Crlf, // CarriageReturn followed by LineFeed
+    LF,   // U+000A -- LineFeed
+    CR,   // U+000D -- CarriageReturn
+    Nel,  // U+0085 -- NextLine
+    LS,   // U+2028 -- Line Separator
+    VT,   // U+000B -- VerticalTab
+    FF,   // U+000C -- FormFeed
+    PS,   // U+2029 -- ParagraphSeparator
+}
+
+pub fn rope_slice_to_line_ending(g: &RopeSlice) -> Option<LineEnding> {
+    if let Some(text) = g.as_str() {
+        str_to_line_ending(text)
+    } else if g == "\u{000D}\u{000A}" {
```

Comment:
@cessen Given that ropey ensures \r\n is never split through two chunks, I think this case isn't necessary?

Reply:
Correct.  That was the simplification I was alluding to in one of my other comments here.  If `as_str()` returns `None`, then there's no way it can be a well-formed line-ending grapheme.